Refactor/Replace Compression & Decompression Filters

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | not yet
| BC Break      | yes

### Description

The old `Compress` and `Decompress` filters were a mess. Different adapters had different or unknown capabilities. Disregarding the now removed formats such as rar and lzf, the Tar and Zip adapters cannot simply compress arbitrary strings but can of course be used to compress and decompress files/archives, and the BZ2 and GZ adapters aren't suited to working with archives but are capable of compressing arbitary strings.

Instead of attempting to do all of these things with all the adapters, I've split the adapters into 2 interfaces:

- `ArchiveAdapterInterface`
- `StringCompressionAdapterInterface`

and replaced the `Compress` and `Decompress` filters with

- `CompressString`
- `CompressToArchive`
- `DecompressString`
- `DecompressArchive`

The file related filter will now accept paths, directories, PSR uploads and PHP `$_FILES` arrays.

The tools added here will also be useful for the other file related filters.

I've introduced the concept of an `ArchiveAdapterResolver`, shipping concrete implementations by default, that first scan the mime-type of the file to see if we have a matching adapter and then guess the filetype from the filename extension.

TODO:

- [ ] Documentation and migration
- [x] A PR for the deprecations

gsteel committed:
Add String De/Compression Filters

Adds a new interface for string compression adapters with implementations for GZ and BZ2 along with dedicated string compression and decompression filters.

Further refactoring of related filters will include more adapters and filters for dealing with files explicitly

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-work the Tar and Zip adapters to implement a new File specific interface

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Introduce `FileInformation` to abstract file handling for paths, psr or php uploads

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Extract tmp directory cleanup to utility

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add compress to archive filter

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove legacy `Compress` and `Decompress` filters

Also enables removal of the legacy `Compression Algorithm` interface and abstract

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Minor SA fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove unused Exception

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove various unused fixtures

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Require minimum recent pear version for compat with 8.2+

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
@froschdesign - I'd appreciate a review on the general direction of this patch when you get the chance, before I do more work on the outstanding items.

Reviews welcome from anyone else too :)

cc @alexmerlin, @ramchale, @Xerkus, @samsonasik

froschdesign commented on src/Compress/Bz2.php:
```diff
@@ -4,183 +4,66 @@
- * @extends AbstractCompressionAlgorithm<Options>
  */
-final class Bz2 extends AbstractCompressionAlgorithm
+final class Bz2 implements StringCompressionAdapterInterface
```
@gsteel 
They are not filters, but adapters for filters. Unfortunately, this is not clear from the name of the class or namespace. Only the interface indicates this.
I think it would be helpful to make this clearer somehow.

gsteel replied:
Considering they are complete rewrites too, I agree that renaming them is probably a good idea 👍

ramchale commented on test/File/FileInformationTest.php:
```diff
@@ -0,0 +1,107 @@
+    {
+        $path = __DIR__ . '/fixtures/File1.txt';
+
+        $upload = new UploadedFile(
+            $path,
+            filesize($path),
+            UPLOAD_ERR_OK,
+            'Foo.txt',
+            'text/plain',
+        );
+
+        $file = FileInformation::factory($upload);
```
If you want to avoid the Laminas\Diactoros\UploadedFile import
```suggestion
                $mockStream = $this->createMock(StreamInterface::class);
        $mockStream->expects(self::once())->method('getMetadata')->with('uri')->willReturn($path);

        $mockFileUpload = $this->createMock(UploadedFileInterface::class);
        $mockFileUpload->expects(self::once())->method('getStream')->willReturn($mockStream);
        $mockFileUpload->expects(self::once())->method('getClientFilename')->willReturn('Foo.txt');
        $mockFileUpload->expects(self::once())->method('getClientMediaType')->willReturn('text/plain');
        
        $file = FileInformation::factory($mockFileUpload);
```

froschdesign replied:
The `UploadedFile` is simple and therefore I would not use a mock here.
And it's also much easier to read.

ramchale replied:
That's fair. The mocking is little more complex. laminas/laminas-diactoros is only being imported for two tests, so this is just an alternative approach for if there's more value in not introducing a new dependency, but it is a tradeoff. I'm not precious either way

ramchale commented on test/CompressToArchiveTest.php:
```diff
@@ -0,0 +1,196 @@
+    {
+        $filter = new CompressToArchive($options);
+        $path   = __DIR__ . '/Compress/fixtures/directory-to-compress/File1.txt';
+        $upload = new UploadedFile(
+            $path,
+            filesize($path),
+            UPLOAD_ERR_OK,
+            'Foo.txt',
+            'text/plain',
+        );
+
+        self::assertFileDoesNotExist($options['archive']);
+        $result = $filter->__invoke($upload);
```
If you want to avoid the Laminas\Diactoros\UploadedFile import
```suggestion
        $mockStream = $this->createMock(StreamInterface::class);
        $mockStream->expects(self::once())->method('getMetadata')->willReturn($path);

        $mockFileUpload = $this->createMock(UploadedFileInterface::class);
        $mockFileUpload->expects(self::once())->method('getStream')->willReturn($mockStream);

        self::assertFileDoesNotExist($options['archive']);
        $result = $filter->__invoke($mockFileUpload);
```

ramchale reviewed (commented):
Looks really good

Added a couple of suggestions that I think mitigate the need to add the dev dependency for laminas/laminas-diactoros, but it depends how much mocking you want to use.

gsteel committed:
Rename compression adapters

- Bz2 -> Bz2Adapter
- Gz -> GzAdapter
- Tar -> TarAdapter
- Zip -> ZipAdapter

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add file-name matcher utility to determine the correct adapter to use

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Use `Archive_Tar`'s automatic archive format detection

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Also accept bz2 compressed tar archives

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Ensure tests fail on PHPUnit deprecations

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Initial filter for decompressing archives

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Test coverage improvements

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add mime-type detection capabilities to `FileInformation`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor the matcher to accept a `FileInformation` instance.

This is so that matchers have more information to use to determine an appropriate adapter

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Introduce MimeType archive matcher and aggregate matcher

In order to recognise the type of archive when it's an uploaded file, and does not have a file name extension, the mime type matcher can be used to guess the mime type and match to the appropriate adapter.

The aggregate matcher allows chaining of matchers so that multiple strategies can be attempted.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove immutable annotation - it's not actually true and lots of impure methods are in use.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Suppress unused method issues at config file level

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Minor CS fix

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Baseline non-use of AbstractFilter::getOptions

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve naming of archive adapters

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Cover mismatched adapters during string de/compression

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update documentation with new compression related filters

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Ignore documentation related tooling and files

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Minor linting fix

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on .gitignore:
```diff
@@ -6,3 +6,5 @@
 /laminas-mkdoc-theme/
 /phpunit.xml
 /vendor/
+/documentation-theme/
+.markdownlint.json
```
Why are these entries needed?

gsteel replied:
These are for local development - If you're checking the docs locally, it'll prevent you from accidentally committing those files. Want a revert?

froschdesign replied:
I have no problem with the entries, I asked because we did not have them before.

froschdesign reviewed (approved):
Great improvements!

gsteel commented on src/Compress/AggregateArchiveAdapterResolver.php:
```diff
@@ -0,0 +1,34 @@
+use function array_values;
+use function sprintf;
+
+final class AggregateArchiveAdapterResolver implements ArchiveAdapterResolverInterface
```
This "Aggregate Adapter Resolver" can be used to compose a chain of concrete adapter resolvers. The first successful resolver wins.

gsteel commented on src/Compress/Bz2Adapter.php:
```diff
@@ -0,0 +1,69 @@
+ *     blocksize?: int<1, 9>|null,
+ * }
+ */
+final class Bz2Adapter implements StringCompressionAdapterInterface
```
The BZ2 adapter now only supports string content.

In future, it will be possible to introduce a `FileCompressionInterface` that this adapter _could_ also implement so that we can filter an arbitrary string, or a single un-compressed file to `somefile.bz2`

gsteel commented on src/Compress/FileExtensionArchiveAdapterResolver.php:
```diff
@@ -0,0 +1,34 @@
+
+use const PATHINFO_EXTENSION;
+
+final class FileExtensionArchiveAdapterResolver implements ArchiveAdapterResolverInterface
```
Simply matches `whatever.zip` to use the Zip archive adapter based on filename extension. By default it is composed last in the chain, preferring mime-magic based mime detection.

gsteel commented on src/Compress/GzAdapter.php:
```diff
@@ -0,0 +1,82 @@
+ *     mode?: 'deflate'|'compress',
+ * }
+ */
+final class GzAdapter implements StringCompressionAdapterInterface
```
As per the BZ2 adapter, the GZ adapter now only supports string content.

In future, it will be possible to introduce a `FileCompressionInterface` that this adapter could also implement so that we can filter an arbitrary string, or a single un-compressed file to `somefile.gz`

gsteel commented on src/Compress/MimeTypeArchiveAdapterResolver.php:
```diff
@@ -0,0 +1,24 @@
+
+use function sprintf;
+
+final class MimeTypeArchiveAdapterResolver implements ArchiveAdapterResolverInterface
```
Map the detected mime-type of a file to an archive adapter - the default method of adapter selection

gsteel commented on src/File/FileInformation.php:
```diff
@@ -0,0 +1,116 @@
+
+use const FILEINFO_MIME_TYPE;
+
+final class FileInformation
```
General purpose file information utility. This would have been marked as `@internal`, but we use it to match unknown files to archive adapters in the interface. Marking it as `@internal` would have been a pain in the ass for anyone who wants to implement a different detection strategy

gsteel committed:
Add `DecompressArchive` filter to the plugin manager

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel reviewed (commented):
Adding some notes to the PR for posterity!
